### ISSUE TITLE
bind_to_device flag doesn't work with gen_tcp_socket

### ISSUE THREAD

**Issue by abbradar** [erlang/otp#6509]
**Describe the bug**

Option `bind_to_device` doesn't work with `{inet_backend, socket}`.

**To Reproduce**

```
6> {Code, Port} = gen_tcp:connect({8, 8, 8, 8}, 53, [{inet_backend, socket}, {bind_to_device, <<"wlp2s0">>}]).
** exception error: no match of right hand side value 
                    {error,{invalid,{socket_option,{socket,bindtodevice},
                                                   <<"wlp2s0">>}}}
7> 
7> {Code, Port} = gen_tcp:connect({8, 8, 8, 8}, 53, [{inet_backend, socket}, {bind_to_device, "wlp2s0"}]).    
** exception exit: badarg
     in function  gen_tcp_socket:connect_lookup/4 (gen_tcp_socket.erl, line 137)
     in call from gen_tcp_socket:connect/4 (gen_tcp_socket.erl, line 103)
8> 
```

Notice that I try both ways of passing the interface name -- charlist and binary. This is related to the explanation below.

**Expected behavior**
Port is successfully created.

**Affected versions**
OTP 24, OTP 25.

**Additional context**

This happens because `gen_tcp_socket` [uses](https://github.com/erlang/otp/blob/d95d79d1e363879462d9775316896c02b59771ee/lib/kernel/src/gen_tcp_socket.erl#L119) `inet.connect_options` to validate options list, It, in turn, [uses](https://github.com/erlang/otp/blob/d95d79d1e363879462d9775316896c02b59771ee/lib/kernel/src/inet.erl#L1382) `prim_inet:is_sockopt_eval` to validate individual options, which expects that the interface name is passed as a binary. Meanwhile, the actual `nif_setopt` from `prim_socket` [expects](https://github.com/erlang/otp/blob/d95d79d1e363879462d9775316896c02b59771ee/erts/emulator/nifs/common/prim_socket_nif.c#L11333) interface name as a charlist. Therefore I expect any string socket option to be broken with `gen_tcp_socket`, not just `bind_to_device`.

I'm not sure which place is the best for addressing this issue -- should `gen_tcp_socket` have its own validation step, or should we instead allow binaries as strings for `prim_socket:nif_setopt`? I'm weakly in favor of the option (2) as the one easier to implement. In fact I have a patch ready that allows binaries to be accepted for string-expecting socket options, fixing this issue.

**Comment by bmk** [comment]
'socket' does support bindtodevice (as the option is called in 'socket').
But, the option types in prim_inet and 'socket' are mutually exclusive: 
prim_inet: Require that the device is of type 'binary()' (<<"wlp2s0">>)
socket:      Require that the device type is of type 'string()' ("wlp2s0")

So the first fails in the socket nif and the second prim_inet validation...

https://github.com/bmk/otp/tree/bmk/kernel/bind_to_device_value_type_conflict/OTP-18357

**Comment by bmk** [comment]
Prepared for patch och OTP 24 and 25.

**Comment by bmk** [comment]
The 24 patch has been released; 24.3.4.7

**Comment by bmk** [comment]
The 25 patch has now been release; 25.2.1
